external-crates/move/crates/move-compiler/src/naming/translate.rs

```diff
@@ -501,45 +567,131 @@ impl<'env> Context<'env> {
         if let Some((id, block_type)) = id_opt {
-            if *block_type == expected_block_type {
-                let nvar_ = N::Var_ {
-                    name,
-                    id: *id,
-                    color: 0,
-                };
-                Some(BlockLabel(sp(loc, nvar_)))
+            let block_type = *block_type;
+            if block_type.is_acceptable_usage(usage) {
+                Some(block_label(
+                    loc,
+                    NominalBlockUsage::Return,
+                    Some(name),
+                    *id,
+                    block_type,
+                ))
             } else {
-                let msg = format!(
-                    "Invalid usage of '{}' with a {} block label",
-                    verb, block_type
-                );
+                let msg = format!("Invalid usage of '{usage}' with a {block_type} block label",);
                 let mut diag = diag!(NameResolution::InvalidLabel, (loc, msg));
-                match expected_block_type {
-                    NominalBlockType::Loop => {
-                        diag.add_note("Loop labels may only be used with 'break' and 'continue', not 'return'");
+                diag.add_note(match block_type {
+                    NominalBlockType::Loop(_) => {
+                        "Loop labels may only be used with 'break' and 'continue', \
+                        not 'return'"
                     }
                     NominalBlockType::Block => {
-                        diag.add_note("Named block labels may only be used with 'return', not 'break' or 'continue'.");
+                        "Named block labels may only be used with 'return', \
+                        not 'break' or 'continue'."
                     }
-                }
+                    NominalBlockType::Lambda => {
+                        "Lambda block labels may only be used with 'return' or 'break', \
+                        not 'continue'."
+                    }
+                });
                 self.env.add_diag(diag);
                 None
             }
         } else {
-            let msg = format!("Invalid {}. Unbound label '{}", verb, name);
+            let msg = format!("Invalid {usage}. Unbound label '{name}");
             self.env
-                .add_diag(diag!(NameResolution::UnboundVariable, (loc, msg)));
+                .add_diag(diag!(NameResolution::UnboundLabel, (loc, msg)));
             None
         }
     }
 
     fn exit_nominal_block(&mut self, loc: Loc) -> BlockLabel {
-        let (name, id, _) = self.nominal_blocks.pop().unwrap();
-        let nvar_ = N::Var_ { name, id, color: 0 };
-        BlockLabel(sp(loc, nvar_))
+        let (_, name, id, name_type) = self.nominal_blocks.pop().unwrap();
+        assert_ne!(name_type, NominalBlockType::Lambda);
+        block_label(
+            loc,
+            /* unused */ NominalBlockUsage::Return,
+            name,
+            id,
+            name_type,
+        )
```

Comment:
```suggestion
        let (_, _, label, name_type) = self.nominal_blocks.pop().unwrap();
        assert_ne!(name_type, NominalBlockType::Lambda);
        label
```

Reply:
Going to do a bit of refactoring here. Should be a lot cleaner